[DTensor][Partial] fixing .item() when dtensor is Partial

anshul-si opened the pull request:
**Summary:** Currently, when users call .item() on a Partial DTensor, each rank returns the local tensor value instead of the global tensor value. The root cause is for aten._local_scalar_dense.default, we return OutputSharding with the original schema, which would be correct for Shard(n) and Replicate(). However, for Partial, we should return a redistribution schema that has Replicate placements and sets needs_redistribution to True. This will cause a redistribution to happen before .item() executes.

**Test Cases**
1. pytest test/distributed/tensor/test_tensor_ops.py -k test_item

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #167598



cc @H-Huang @awgu @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @pragupta @msaroufim @dcci @tianyu-l @XilunWu @SherlockNoMad @ppwwyyxx

anshul-si committed:
[DTensor][Partial] fixing .item() when dtensor is Partial #167595

[ghstack-poisoned]

anshul-si committed:
Update on "[DTensor][Partial] fixing .item() when dtensor is Partial #167595"




cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta msaroufim dcci tianyu-l XilunWu SherlockNoMad

[ghstack-poisoned]

anshul-si committed:
Update on "[DTensor][Partial] fixing .item() when dtensor is Partial #167595"

**Summary:** Currently, when users call .item() on a Partial DTensor, each rank returns the local tensor value instead of the global tensor value. The root cause is for aten._local_scalar_dense.default, we return OutputSharding with the original schema, which would be correct for Shard(n) and Replicate(). However, for Partial, we should return a redistribution schema that has Replicate placements and sets needs_redistribution to True. This will cause a redistribution to happen before .item() executes.

**Test Cases**
1. pytest test/distributed/tensor/test_tensor_ops.py -k test_item




cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta msaroufim dcci tianyu-l XilunWu SherlockNoMad

[ghstack-poisoned]

wconstab commented on torch/distributed/tensor/_sharding_prop.py:
```diff
@@ -366,6 +367,41 @@ def propagate_op_sharding_non_cached(self, op_schema: OpSchema) -> OutputShardin
+                        has_partial = True
+                        break
+
+                if has_partial:
```
something like `if any(p.is_partial() for p in arg.placements)` should work instead of the block above?

wconstab commented on torch/distributed/tensor/_sharding_prop.py:
```diff
@@ -366,6 +367,41 @@ def propagate_op_sharding_non_cached(self, op_schema: OpSchema) -> OutputShardin
+                    return OutputSharding(
+                        None,
+                        suggestion_schema,
+                        # needs_redistribute=True,
```
clean this up? (is this being set to true automatically in the rewrap_schema thing above, when there is a partial?

wconstab commented on torch/distributed/tensor/_sharding_prop.py:
```diff
@@ -366,6 +367,41 @@ def propagate_op_sharding_non_cached(self, op_schema: OpSchema) -> OutputShardin
+                        (replicate_spec,),
+                        {},
+                    )
+                    suggestion_schema._inplace_rewrap_schema_suggestion(op_schema)
```
i'm not actually familiar with `_inplace_rewrap_schema_suggestion`, can you educate me on what its for?  is it just to avoid mutating 'op_schema' yourself?

anshul-si replied:
_inplace_rewrap_schema_suggestion is used in instances where there are non-DTensor args and kwargs from the original schema and copy it over to the new schema. In this case, I can get rid of it since since aten._local_scalar_dense.default takes in only one arguement, which has to be a DTensor, but I added it in as a sanity check / following same format as below

wconstab reviewed (commented):
the comments on the test file were just FYI, no need to fix in this PR.

i did want to get the question on the sharding prop file answered before stamping though.

anshul-si committed:
Update on "[DTensor][Partial] fixing .item() when dtensor is Partial "

**Summary:** Currently, when users call .item() on a Partial DTensor, each rank returns the local tensor value instead of the global tensor value. The root cause is for aten._local_scalar_dense.default, we return OutputSharding with the original schema, which would be correct for Shard(n) and Replicate(). However, for Partial, we should return a redistribution schema that has Replicate placements and sets needs_redistribution to True. This will cause a redistribution to happen before .item() executes.

**Test Cases**
1. pytest test/distributed/tensor/test_tensor_ops.py -k test_item




cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta msaroufim dcci tianyu-l XilunWu SherlockNoMad

[ghstack-poisoned]

anshul-si committed:
Update on "[DTensor][Partial] fixing .item() when dtensor is Partial "

**Summary:** Currently, when users call .item() on a Partial DTensor, each rank returns the local tensor value instead of the global tensor value. The root cause is for aten._local_scalar_dense.default, we return OutputSharding with the original schema, which would be correct for Shard(n) and Replicate(). However, for Partial, we should return a redistribution schema that has Replicate placements and sets needs_redistribution to True. This will cause a redistribution to happen before .item() executes.

**Test Cases**
1. pytest test/distributed/tensor/test_tensor_ops.py -k test_item




cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta msaroufim dcci tianyu-l XilunWu SherlockNoMad

[ghstack-poisoned]

anshul-si committed:
Update on "[DTensor][Partial] fixing .item() when dtensor is Partial "

**Summary:** Currently, when users call .item() on a Partial DTensor, each rank returns the local tensor value instead of the global tensor value. The root cause is for aten._local_scalar_dense.default, we return OutputSharding with the original schema, which would be correct for Shard(n) and Replicate(). However, for Partial, we should return a redistribution schema that has Replicate placements and sets needs_redistribution to True. This will cause a redistribution to happen before .item() executes.

**Test Cases**
1. pytest test/distributed/tensor/test_tensor_ops.py -k test_item




cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta msaroufim dcci tianyu-l XilunWu SherlockNoMad

[ghstack-poisoned]

anshul-si committed:
Update on "[DTensor][Partial] fixing .item() when dtensor is Partial "

**Summary:** Currently, when users call .item() on a Partial DTensor, each rank returns the local tensor value instead of the global tensor value. The root cause is for aten._local_scalar_dense.default, we return OutputSharding with the original schema, which would be correct for Shard(n) and Replicate(). However, for Partial, we should return a redistribution schema that has Replicate placements and sets needs_redistribution to True. This will cause a redistribution to happen before .item() executes.

**Test Cases**
1. pytest test/distributed/tensor/test_tensor_ops.py -k test_item




cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta msaroufim dcci tianyu-l XilunWu SherlockNoMad

[ghstack-poisoned]

ezyang reviewed (changes requested):
This seems overly complicated for what it should do, and it ALSO silently changes the behavior of item() on Partial, when I think what we need to do is first switch it to error, prop it through fbcode, and THEN switch it to the new semantics

anshul-si committed:
Update on "[DTensor][Partial] fixing .item() when dtensor is Partial "

**Summary:** Currently, when users call .item() on a Partial DTensor, each rank returns the local tensor value instead of the global tensor value. The root cause is for aten._local_scalar_dense.default, we return OutputSharding with the original schema, which would be correct for Shard(n) and Replicate(). However, for Partial, we should return a redistribution schema that has Replicate placements and sets needs_redistribution to True. This will cause a redistribution to happen before .item() executes.

**Test Cases**
1. pytest test/distributed/tensor/test_tensor_ops.py -k test_item




cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta msaroufim dcci tianyu-l XilunWu SherlockNoMad

[ghstack-poisoned]

wconstab reviewed (commented):
@ezyang you mean to convert this to a loud error first, and see what breaks, so we can go tell people that they have been experiencing silent failures?

do we then need to land this fbcode-first and change any code that was raising the error to manually avoid the error by doing an explicit redistribute before calling .item?  

and then finally land the PR that supports this?

ezyang commented:
Yes. and you may need to JK flag it too if someone asks you to fix their job without having to redo a code push.

anshul-si commented:
https://github.com/pytorch/pytorch/pull/168030 error message
